### PR TITLE
Various core and test changes

### PR DESCRIPTION
**Pull request by tybug**
Split from #3818. I could split this further if desired, with all the `draw_bits` changes in a single PR.

Changes:
* [move cu.choice to ConjectureData](https://github.com/HypothesisWorks/hypothesis/commit/0690201688cc7024ddebc70ef237791d5cbe5e1c)
* [improve minimal readability with nonlocal](https://github.com/HypothesisWorks/hypothesis/commit/f6b7349c9520b3c08baf698084541b08d9db9ebc)
* migrate away from `draw_bits` in various test files in preparation for its removal. Or at least, taking a back seat relative to the ir.

There were some test changes where the extra boolean draw from > 24 bit integers caused the test to fail.

https://github.com/HypothesisWorks/hypothesis/blob/980042e4018069cf24f6c161434378a353586a76/hypothesis-python/src/hypothesis/internal/conjecture/data.py#L1260-L1262

I mostly remedied this by reducing the bit size down to 20 for these tests. It does make me wonder if we're doing the appropriate thing by biasing the distribution in this way at the ir level, rather than at the strategy level (`integers()`).

### CODE REVIEW

**Comment by tybug** [comment]
Looks like `test_can_learn_to_normalize_the_unnormalized` became flaky due to this ([e.g.](https://github.com/HypothesisWorks/hypothesis/actions/runs/7480537250/job/20360179147?pr=3835)). Will look into it.

**Comment by Zac-HD** [comment]
> I mostly remedied this by reducing the bit size down to 20 for these tests. It does make me wonder if we're doing the appropriate thing by biasing the distribution in this way at the ir level, rather than at the strategy level (`integers()`).

Hmm, good question!  IIRC we used 2**24 as a special value because it ensured that unicode characters were chosen without (additional) bias.  I'd probably leave it alone for now, but we should do an audit of this kind of heuristic once we're using the IR everywhere - there are probably a bunch we'll be able to clean up then.

------------------

General thoughts on this PR: looks good to me.  Two approaches you can take to this kind of pulled-out stage are to (1) defer anything that we can't merge ~immediately to a future PR, or (2) keep working on it here until it passes to make future PRs easier.  I generally favor a bit of (2) followed by (1) if I'm feeling stuck or just slowed down; it's basically the same question as pulling out part in the first place 😁

**Comment by tybug** [comment]
that approach makes sense 👍. 

In this case I was able to figure out `test_can_learn_to_normalize_the_unnormalized`. It was a rare case where a test is failing because we did too good of a job 😄. `dfa.normalize` only adds a new normalization if the shrinking differs within 100 calls. With `draw_bits`, the shrinking differed within that time frame. After using the IR, the shrinking was *usually* consistent (hence the flakiness). Which means we were doing a better job shrinking to a normalized example with the IR than without!

The solution? Make the property even harder to shrink to a normalized value.

**Review by Zac-HD — APPROVED** [review]
🎉

File: hypothesis-python/tests/conjecture/test_optimiser.py [hunk 1]
```diff
@@ -134,7 +134,10 @@ def test_targeting_can_drive_length_very_high():
 
         def test(data):
             count = 0
-            while data.draw_bits(2) == 3:
+            # TODO this test fails with data.draw_boolean(0.25). Does the hill
+            # climbing optimizer just not like the bit representation of boolean
+            # draws, or do we have a deeper bug here?
+            while data.draw_integer(0, 3) == 3:
```
[the remark this answers is not in the document]
Comment (tybug) [reply hypothesis-python/tests/conjecture/test_optimiser.py:140]: left a task in the pr description 👍 (less likely for me to lose it than a comment, and github doesn't easily let me leave reviews on unrelated files.)

**Pull request merged**